

<!doctype html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Tutorial &#8212; mpathic 0.0.1 documentation</title>
    <link rel="stylesheet" href="_static/bizstyle.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/overrides.css" type="text/css" />
    <script type="text/javascript" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/bizstyle.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Examples" href="examples.html" />
    <link rel="prev" title="MPAthic" href="index.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <!--[if lt IE 9]>
    <script type="text/javascript" src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="examples.html" title="Examples"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="MPAthic"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">mpathic 0.0.1 documentation</a> &#187;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Tutorial</a><ul>
<li><a class="reference internal" href="#simulating-data">Simulating Data</a></li>
<li><a class="reference internal" href="#computing-profiles">Computing Profiles</a></li>
<li><a class="reference internal" href="#quantitative-modeling">Quantitative Modeling</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">MPAthic</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="examples.html"
                        title="next chapter">Examples</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/tutorial.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="tutorial">
<h1>Tutorial<a class="headerlink" href="#tutorial" title="Permalink to this headline">¶</a></h1>
<p>Import the MPAthic package as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">mpathic</span> <span class="k">as</span> <span class="nn">mpa</span>
</pre></div>
</div>
<div class="section" id="simulating-data">
<h2>Simulating Data<a class="headerlink" href="#simulating-data" title="Permalink to this headline">¶</a></h2>
<p>We begin by simulating a library of variant CRP binding sites. We can use the <a class="reference internal" href="simulate_library.html"><span class="doc">mpa.SimulateLibrary</span></a> class to
create a library of random mutants from an initial wildtype sequence and mutation rate:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sim_library</span> <span class="o">=</span> <span class="n">mpa</span><span class="o">.</span><span class="n">SimulateLibrary</span><span class="p">(</span><span class="n">wtseq</span><span class="o">=</span><span class="s2">&quot;TAATGTGAGTTAGCTCACTCAT&quot;</span><span class="p">,</span> <span class="n">mutrate</span><span class="o">=</span><span class="mf">0.24</span><span class="p">)</span>
<span class="n">sim_library</span><span class="o">.</span><span class="n">output_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<p>The <cite>output_df</cite> attribute of the <code class="docutils literal notranslate"><span class="pre">sim_library</span></code> class looks like the dataframe below</p>
<table border="1" class="docutils">
<colgroup>
<col width="38%" />
<col width="63%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">ct</th>
<th class="head">seq</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>21</td>
<td>TAATGTGAGTTAGCTCACTCAT</td>
</tr>
<tr class="row-odd"><td>7</td>
<td>TAATGTGAGTTAGCTAACTCAT</td>
</tr>
<tr class="row-even"><td>6</td>
<td>TAATGTGAGTTAGCTCACTCAA</td>
</tr>
</tbody>
</table>
<p>⋮</p>
<table border="1" class="docutils">
<colgroup>
<col width="38%" />
<col width="63%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>1</td>
<td>TAATGTGTGTTCGCTCATCCAT</td>
</tr>
</tbody>
</table>
<p>In general, MPAthic datasets are pandas dataframes, comprising of columns of counts and sequence values. To simulate
a Sort-Seq experiment (<a class="footnote-reference" href="#kinney2010" id="id1">[1]</a>), we use the <a class="reference internal" href="simulate_sort.html"><span class="doc">mpa.SimulateSort</span></a> class. This class requires a dataset input
and a model dataframe input. We first import these inputs using <code class="docutils literal notranslate"><span class="pre">io</span></code> module provided with the MPAthic package:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load dataset and model dataframes</span>
<span class="n">dataset_df</span> <span class="o">=</span> <span class="n">mpa</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="s1">&#39;sort_seq_data.txt&#39;</span><span class="p">)</span>
<span class="n">model_df</span> <span class="o">=</span> <span class="n">mpa</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">load_model</span><span class="p">(</span><span class="s1">&#39;crp_model.txt&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Next, we call the <code class="docutils literal notranslate"><span class="pre">SimulateSort</span></code> class as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Simulate a Sort-Seq experiment</span>
<span class="n">sim_sort</span> <span class="o">=</span> <span class="n">mpa</span><span class="o">.</span><span class="n">SimulateSort</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">dataset_df</span><span class="p">,</span><span class="n">mp</span><span class="o">=</span><span class="n">model_df</span><span class="p">)</span>
<span class="n">sim_sort</span><span class="o">.</span><span class="n">output_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<p>The head of the output dataframe looks like</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="10%" />
<col width="10%" />
<col width="10%" />
<col width="10%" />
<col width="10%" />
<col width="41%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">ct</th>
<th class="head">ct_0</th>
<th class="head">ct_1</th>
<th class="head">ct_2</th>
<th class="head">ct_3</th>
<th class="head">ct_4</th>
<th class="head">seq</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>4</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>AAAAAAGGTGAGTTAGCTAACT</td>
</tr>
<tr class="row-odd"><td>3</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>AAAAAATATAAGTTAGCTCGCT</td>
</tr>
<tr class="row-even"><td>4</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>0</td>
<td>AAAAAATATGATTTAGCTGACT</td>
</tr>
<tr class="row-odd"><td>3</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>AAAAAATGTCAGTTAGCTCACT</td>
</tr>
<tr class="row-even"><td>4</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>AAAAAATGTGAATTATCGCACT</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="computing-profiles">
<h2>Computing Profiles<a class="headerlink" href="#computing-profiles" title="Permalink to this headline">¶</a></h2>
<p>It is often useful to compute the mutation rate within a set of sequences, e.g., in order to validate the
composition of a library. This can be accomplished using the <a class="reference internal" href="profile_mutrate.html"><span class="doc">mpa.ProfileMutrate</span></a> class as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">profile_mut</span> <span class="o">=</span> <span class="n">mpa</span><span class="o">.</span><span class="n">ProfileMut</span><span class="p">(</span><span class="n">dataset_df</span> <span class="o">=</span> <span class="n">dataset_df</span><span class="p">)</span>
<span class="n">profile_mut</span><span class="o">.</span><span class="n">mut_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<p>The mutation rate at each position within the sequences looks like</p>
<table border="1" class="docutils">
<colgroup>
<col width="26%" />
<col width="21%" />
<col width="53%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">pos</th>
<th class="head">wt</th>
<th class="head">mut</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>0</td>
<td>A</td>
<td>0</td>
</tr>
<tr class="row-odd"><td>1</td>
<td>A</td>
<td>0</td>
</tr>
<tr class="row-even"><td>2</td>
<td>A</td>
<td>0.33871</td>
</tr>
<tr class="row-odd"><td>3</td>
<td>T</td>
<td>0.127566</td>
</tr>
<tr class="row-even"><td>4</td>
<td>A</td>
<td>0.082111</td>
</tr>
</tbody>
</table>
<p>To view the frequency of occurrence for every base at each position, use the <a class="reference internal" href="profile_freq.html"><span class="doc">mpa.ProfileFreq</span></a> class:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">profile_freq</span> <span class="o">=</span> <span class="n">mpa</span><span class="o">.</span><span class="n">ProfileFreq</span><span class="p">(</span><span class="n">dataset_df</span> <span class="o">=</span> <span class="n">dataset_df</span><span class="p">)</span>
<span class="n">profile_freq</span><span class="o">.</span><span class="n">freq_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<table border="1" class="docutils">
<colgroup>
<col width="11%" />
<col width="22%" />
<col width="22%" />
<col width="22%" />
<col width="22%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">pos</th>
<th class="head">freq_A</th>
<th class="head">freq_C</th>
<th class="head">freq_G</th>
<th class="head">freq_T</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>0</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="row-odd"><td>1</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="row-even"><td>2</td>
<td>0.66129</td>
<td>0.33871</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="row-odd"><td>3</td>
<td>0.043988</td>
<td>0.042522</td>
<td>0.041056</td>
<td>0.872434</td>
</tr>
<tr class="row-even"><td>4</td>
<td>0.917889</td>
<td>0.019062</td>
<td>0.02566</td>
<td>0.03739</td>
</tr>
</tbody>
</table>
<p>Information pro les (also called “information footprints”) provide a particularly useful way to identify
functional positions within a sequence. These pro les list, for each position in a sequence, the mutual
information between the character at that position and the bin in which a sequence is found. Unlike mutation
and frequency profiles, which require sequence counts for a single bin only, information profiles are
computed from full datasets, and can be accomplished using the <a class="reference internal" href="profile_info.html"><span class="doc">mpa.ProfileInfo</span></a> class as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">profile_info</span> <span class="o">=</span> <span class="n">mpa</span><span class="o">.</span><span class="n">ProfileInfo</span><span class="p">(</span><span class="n">dataset_df</span> <span class="o">=</span> <span class="n">dataset_df</span><span class="p">)</span>
<span class="n">profile_info</span><span class="o">.</span><span class="n">info_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<table border="1" class="docutils">
<colgroup>
<col width="33%" />
<col width="67%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">pos</th>
<th class="head">info</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>0</td>
<td>0.000077</td>
</tr>
<tr class="row-odd"><td>1</td>
<td>0.000077</td>
</tr>
<tr class="row-even"><td>2</td>
<td>0.008357</td>
</tr>
<tr class="row-odd"><td>3</td>
<td>0.008743</td>
</tr>
<tr class="row-even"><td>4</td>
<td>0.013745</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="quantitative-modeling">
<h2>Quantitative Modeling<a class="headerlink" href="#quantitative-modeling" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference internal" href="learn_model.html"><span class="doc">mpa.LearnModel</span></a> class can be used to  fit quantitative models to data:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">learned_model</span> <span class="o">=</span> <span class="n">mpa</span><span class="o">.</span><span class="n">LearnModel</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">dataset_df</span><span class="p">)</span>
<span class="n">learned_model</span><span class="o">.</span><span class="n">output_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<table border="1" class="docutils">
<colgroup>
<col width="11%" />
<col width="23%" />
<col width="23%" />
<col width="21%" />
<col width="21%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">pos</th>
<th class="head">val_A</th>
<th class="head">val_C</th>
<th class="head">val_G</th>
<th class="head">val_T</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>0</td>
<td>-0.201587</td>
<td>0.067196</td>
<td>0.067196</td>
<td>0.067196</td>
</tr>
<tr class="row-odd"><td>1</td>
<td>-0.201587</td>
<td>0.067196</td>
<td>0.067196</td>
<td>0.067196</td>
</tr>
<tr class="row-even"><td>2</td>
<td>-0.10637</td>
<td>-0.167351</td>
<td>0.13686</td>
<td>0.13686</td>
</tr>
<tr class="row-odd"><td>3</td>
<td>-0.287282</td>
<td>0.041222</td>
<td>-0.2039</td>
<td>0.44996</td>
</tr>
<tr class="row-even"><td>4</td>
<td>-0.056109</td>
<td>-0.871858</td>
<td>0.344537</td>
<td>0.583429</td>
</tr>
</tbody>
</table>
<p>The purpose of having a quantitative model is to be able to predict the activity
of arbitrary sequences. This basic operation is accomplished using the <a class="reference internal" href="evaluate_model.html"><span class="doc">mpa.EvaluateModel</span></a> class:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">eval_model</span> <span class="o">=</span> <span class="n">mpa</span><span class="o">.</span><span class="n">EvaluateModel</span><span class="p">(</span><span class="n">dataset_df</span> <span class="o">=</span> <span class="n">dataset_df</span><span class="p">,</span> <span class="n">model_df</span> <span class="o">=</span> <span class="n">model_df</span><span class="p">)</span>
<span class="n">eval_model</span><span class="o">.</span><span class="n">out_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="9%" />
<col width="9%" />
<col width="9%" />
<col width="9%" />
<col width="9%" />
<col width="35%" />
<col width="16%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">ct</th>
<th class="head">ct_0</th>
<th class="head">ct_1</th>
<th class="head">ct_2</th>
<th class="head">ct_3</th>
<th class="head">ct_4</th>
<th class="head">seq</th>
<th class="head">val</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>1</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>AAAGGTGAGTTAGCTAACTCAT</td>
<td>0.348108</td>
</tr>
<tr class="row-odd"><td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>AAATATAAGTTAGCTCGCTCAT</td>
<td>-0.248134</td>
</tr>
<tr class="row-even"><td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>0</td>
<td>AAATATGATTTAGCTGACTCAT</td>
<td>0.009507</td>
</tr>
<tr class="row-odd"><td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>AAATGTCAGTTAGCTCACTCAT</td>
<td>0.238852</td>
</tr>
<tr class="row-even"><td>1</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>AAATGTGAATTATCGCACTCAT</td>
<td>-0.112121</td>
</tr>
</tbody>
</table>
<p>Often, it is useful to scan a model over all sequences embedded within larger contigs. To
do this, MPAthic provides the class <a class="reference internal" href="scan_model.html"><span class="doc">mpa.ScanModel</span></a>, which is called as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># get contigs, provided with mpathic</span>
<span class="n">fastafile</span> <span class="o">=</span> <span class="s2">&quot;./mpathic/examples/genome_ecoli_1000lines.fa&quot;</span>
<span class="n">contig</span> <span class="o">=</span> <span class="n">mpa</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">load_contigs_from_fasta</span><span class="p">(</span><span class="n">fastafile</span><span class="p">,</span> <span class="n">model_df</span><span class="p">)</span>

<span class="n">scanned_model</span> <span class="o">=</span> <span class="n">mpa</span><span class="o">.</span><span class="n">ScanModel</span><span class="p">(</span><span class="n">model_df</span> <span class="o">=</span> <span class="n">model_df</span><span class="p">,</span> <span class="n">contigs_list</span> <span class="o">=</span> <span class="n">contigs_list</span><span class="p">)</span>
<span class="n">scanned_model</span><span class="o">.</span><span class="n">sitelist_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<table border="1" class="docutils">
<colgroup>
<col width="4%" />
<col width="15%" />
<col width="36%" />
<col width="10%" />
<col width="10%" />
<col width="7%" />
<col width="16%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">&#160;</th>
<th class="head">val</th>
<th class="head">seq</th>
<th class="head">left</th>
<th class="head">right</th>
<th class="head">ori</th>
<th class="head">contig</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>0</td>
<td>2.040628</td>
<td>GGTCGTTTGCCTGCGCCGTGCA</td>
<td>11710</td>
<td>11731</td>
<td><ul class="first last simple">
<li></li>
</ul>
</td>
<td>MG1655.fa</td>
</tr>
<tr class="row-odd"><td>1</td>
<td>2.00608</td>
<td>GGAAGTCGCCGCCCGCACCGCT</td>
<td>74727</td>
<td>74748</td>
<td><ul class="first last simple">
<li></li>
</ul>
</td>
<td>MG1655.fa</td>
</tr>
<tr class="row-even"><td>2</td>
<td>1.996992</td>
<td>TGGGTGTGGCGCGTGACCTGTT</td>
<td>45329</td>
<td>45350</td>
<td><ul class="first last simple">
<li></li>
</ul>
</td>
<td>MG1655.fa</td>
</tr>
<tr class="row-odd"><td>3</td>
<td>1.920821</td>
<td>GGTATGTGTCGCCAGCCAGGCA</td>
<td>38203</td>
<td>38224</td>
<td><ul class="first last simple">
<li></li>
</ul>
</td>
<td>MG1655.fa</td>
</tr>
<tr class="row-even"><td>4</td>
<td>1.879852</td>
<td>GGTGATTTTGGCGTGGTGGCGT</td>
<td>73077</td>
<td>73098</td>
<td><ul class="first last simple">
<li></li>
</ul>
</td>
<td>MG1655.fa</td>
</tr>
</tbody>
</table>
<p>A good way to assess the quality of a model is to compute its predictive information on a massively
parallel data set. This can be done using the <cite>predictive_info</cite> (need to write this) class:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">predictive_info</span> <span class="o">=</span> <span class="n">mpa</span><span class="o">.</span><span class="n">PredictiveInfo</span><span class="p">(</span><span class="n">data_df</span> <span class="o">=</span> <span class="n">dataset_df</span><span class="p">,</span> <span class="n">model_df</span> <span class="o">=</span> <span class="n">model_df</span><span class="p">,</span><span class="n">start</span><span class="o">=</span><span class="mi">52</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>References</strong></p>
<table class="docutils footnote" frame="void" id="kinney2010" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id1">[1]</a></td><td>Kinney JB, Anand Murugan, Curtis G. Callan Jr., and Edward C. Cox (2010) <a class="reference external" href="http://www.pnas.org/content/107/20/9158">Using deep sequencing to characterize the biophysical mechanism of a transcriptional regulatory sequence.</a> PNAS May 18, 2010. 107 (20) 9158-9163;
<a class="reference download internal" href="_downloads/Kinney2010.pdf" download=""><code class="xref download docutils literal notranslate"><span class="pre">PDF</span></code></a>.</td></tr>
</tbody>
</table>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="examples.html" title="Examples"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="MPAthic"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">mpathic 0.0.1 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2018, Ammar Tareen, William Ireland, Justin Kinney.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.7.4.
    </div>
  </body>
</html>